### PR TITLE
Import lightweight unit tests

### PR DESCRIPTION
**Pull request by muneebahmed10**
*Description of changes:*

Import unit tests for lightweight API from [existing tests](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/libraries/standard/mqtt/test/unit/iot_tests_mqtt_api.c#L2000) (Line 2000 onwards). Does not include tests for functions that are not in v4_beta (`MQTT_SerializeAck()` and `MQTT_SerializePublishHeader()`)

Total coverage:
* Lines: 91% coverage
* Functions: 91% coverage
* Branches: 90% coverage

Tested functions have 100% coverage. Besides altering names and parameters to fit the new API, some changes were made to existing tests to achieve coverage, which are called out in the comments below.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 1]
```diff
@@ -0,0 +1,1198 @@
...
+    status = MQTT_SerializePublish( &publishInfo,
+                                    PACKET_ID,
+                                    remainingLength,
+                                    &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    mqttPacketInfo.type = buffer[ 0 ];
+    mqttPacketInfo.remainingLength = ( size_t ) buffer[ 1 ];
+    mqttPacketInfo.pRemainingData = &buffer[ 2 ];
+    status = MQTT_DeserializePublish( &mqttPacketInfo, &packetIdentifier, &publishInfo );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:1115]: Added test of QoS publish

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 2]
```diff
@@ -0,0 +1,1198 @@
...
+
+    /* Remaining length of 128 needs 2 bytes. */
+    bufPtr = buffer;
+    buffer[ 0 ] = MQTT_PACKET_TYPE_PUBLISH;
+    buffer[ 1 ] = 0x80;
+    buffer[ 2 ] = 0x01;
+    status = MQTT_GetIncomingPacketTypeAndLength( mockReceive, networkContext, &mqttPacket );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    TEST_ASSERT_EQUAL_INT( MQTT_PACKET_TYPE_PUBLISH, mqttPacket.type );
+    TEST_ASSERT_EQUAL_INT( 128, mqttPacket.remainingLength );
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:1146]: Added this test for branch coverage

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 3]
```diff
@@ -0,0 +1,1198 @@
...
+    bufPtr = buffer;
+    buffer[ 0 ] = MQTT_PACKET_TYPE_PUBREL & 0xF0U;
+    status = MQTT_GetIncomingPacketTypeAndLength( mockReceive, networkContext, &mqttPacket );
+    TEST_ASSERT_EQUAL( MQTTBadResponse, status );
+
+    /* Receive type then fail. */
+    bufPtr = buffer;
+    buffer[ 0 ] = MQTT_PACKET_TYPE_PUBREL;
+    status = MQTT_GetIncomingPacketTypeAndLength( mockReceiveSucceedThenFail, networkContext, &mqttPacket );
+    TEST_ASSERT_EQUAL( MQTTBadResponse, status );
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:1197]: Added these tests due to differences between the old network interface and the new transport receive function, as well as branch coverage.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 4]
```diff
@@ -0,0 +1,1198 @@
...
+
+/* Called at the end of the whole suite. */
+int suiteTearDown( int numFailures )
+{
+    return numFailures;
+}
+
+/* ========================================================================== */
+
+static int32_t mockReceive( MQTTNetworkContext_t context,
```
[the remark this answers is not in the document]
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:138]: I don't believe we can, since the receive function is only a function pointer, and not defined in an external header file.

File: libraries/standard/mqtt/utest/mqtt_config.h [hunk 5]
```diff
@@ -16,14 +16,14 @@
...
-#define LIBRARY_LOG_LEVEL    LOG_INFO
+#define LIBRARY_LOG_LEVEL    LOG_NONE
 
 #include "logging_stack.h"
 
 /************ End of logging configuration ****************/
 
 /* Set network context to socket (int). */
-typedef int MQTTNetworkContext_t;
+typedef uint8_t ** MQTTNetworkContext_t;
```
[the remark this answers is not in the document]
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_config.h:26]: Updated

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 6]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    willInfo.topicNameLength = 4;
+    willInfo.pPayload = "testload";
+    willInfo.payloadLength = 8;
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    /* Make sure remaining size returned is 32 = 16 + 16. */
+    TEST_ASSERT_EQUAL_INT( 32, remainingLength );
+    /* Make sure packet size is 34 = 18 + 16. */
+    TEST_ASSERT_EQUAL_INT( 34, packetSize );
+}
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:391]: I see that the paths for pUserName != NULL and pPassword != NULL and connectPacketSize > MQTT_PACKET_CONNECT_MAX_SIZE are covered in the unit tests for MQTT_SerializeConnect. Considering MQTT_GetConnectPacketSize as a unit, I think those cases belong here.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 7]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    connectInfo.pClientIdentifier = CLIENT_IDENTIFIER;
+    connectInfo.clientIdentifierLength = UINT16_MAX;
+    connectInfo.pPassword = "";
+    connectInfo.passwordLength = UINT16_MAX;
+    connectInfo.pUserName = "";
+    connectInfo.userNameLength = UINT16_MAX;
+    willInfo.pTopicName = TEST_TOPIC_NAME;
+    willInfo.topicNameLength = UINT16_MAX;
+    willInfo.payloadLength = UINT16_MAX + 2;
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:428]: This is unit tests for MQTT_GetConnectPacketSize as mentioned in the earlier comment.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 8]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    memset( ( void * ) &connectInfo, 0x0, sizeof( connectInfo ) );
+    connectInfo.pClientIdentifier = "TEST";
+    connectInfo.clientIdentifierLength = 4;
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    /* Make sure buffer has enough space */
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    /* Make sure test succeeds. */
+    status = MQTT_SerializeConnect( &connectInfo, NULL, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:442]: It will be a good addition to verify the serialized buffer as well here. That would be adding value to tests as we will be verifying the if the serialization(The main output of the function MQTT_SerializeConnect) is as expected. The same comment applies to all the success cases of serialization APIs in this file.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 9]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    connectInfo.userNameLength = 4;
+    connectInfo.pPassword = "PASS";
+    connectInfo.passwordLength = 4;
+    status = MQTT_GetConnectPacketSize( &connectInfo, NULL, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, NULL, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+
+    /* Serialize connect with LWT. */
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:456]: please add a description here about which parameter is bad so that it will help readability of the code.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 10]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    willInfo.qos = MQTTQoS1;
+    willInfo.pTopicName = "test";
+    willInfo.topicNameLength = ( uint16_t ) 4;
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    ( void ) memset( &willInfo, 0x00, sizeof( MQTTPublishInfo_t ) );
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:468]: please add a small description in comments for this test.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 11]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    willInfo.qos = MQTTQoS1;
+    willInfo.pPayload = "test";
+    willInfo.payloadLength = ( uint16_t ) strlen( willInfo.pPayload );
+    status = MQTT_GetConnectPacketSize( &connectInfo, &willInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    ( void ) memset( &willInfo, 0x00, sizeof( MQTTPublishInfo_t ) );
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:479]: same comment about a small description in comments.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 12]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    status = MQTT_SerializeConnect( &connectInfo, &willInfo, remainingLength, &fixedBuffer );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+
+    /* For this example, MQTT_GetConnectPacketSize() will return
+     * packetSize = remainingLength +2 (two byte fixed header).
+     * Make sure MQTT_SerializeConnect()
+     * fails  when remaining length is more than packet size. */
+    /* TODO: Since MQTT_SerializeConnect trusts the remaining length to be
+     * accurate, either change the implementation or remove this test, since it
+     * should fail. */
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:515]: What is the plan to address this? I think fixing it in the library is a good solution.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 13]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+ * @brief Tests that MQTT_GetSubscribePacketSize works as intended.
+ */
+void test_MQTT_GetSubscribePacketSize( void )
+{
+    MQTTSubscribeInfo_t subscriptionList;
+    size_t subscriptionCount = 0;
+    size_t remainingLength = 0;
+    size_t packetSize = 0;
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTSubscribeInfo_t fourThousandSubscriptions[ 4096 ] = { 0 };
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:532]: This test will increase the required stack size by at least 4KB * sizeof(MQTTSubscribeInfo_t ).  As long as unit tests are not targeted to run on targets that are memory constrained, this should be ok.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 8 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:442]: Agree, although I think that needs to be part of a separate PR (since this PR deals mainly with importing existing tests and verifying the buffer would require adding some new code)

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 14]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+                                      remainingLength,
+                                      &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    status = MQTT_SerializeSubscribe( &subscriptionList,
+                                      subscriptionCount,
+                                      PACKET_ID,
+                                      remainingLength,
+                                      NULL );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:687]: Don't we have 2 more paths to cover for invalid params? subscriptionCount == 0U and packetSize > pBuffer->size

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 15]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, status );
+
+    /* Good case succeeds. */
+    publishInfo.pTopicName = "/test/topic";
+    publishInfo.topicNameLength = sizeof( "/test/topic" );
+    publishInfo.pPayload = "";
+    publishInfo.payloadLength = 0;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:868]: Don't we have another branch to cover here "if( pPublishInfo->qos > MQTTQoS0 )" for calculating the packet size?

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 14 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:687]: subscriptionCount == 0 is handled in line 704. packetSize > pBuffer->size is handled in line 726, though I think that's only because the buffer is too small. We can make that a separate test where we more explicitly set the buffer size to a low value

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 15 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:868]: That's already handled when testing MQTT_SerializePublish, though I can copy that code over here too if that's desirable

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 16]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    /* Make sure buffer has enough space */
+    TEST_ASSERT_GREATER_OR_EQUAL( packetSize, bufferSize );
+    status = MQTT_SerializePublish( &publishInfo,
+                                    PACKET_ID,
+                                    remainingLength,
+                                    &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+}
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:974]: How about success cases for QoS1 and QoS0?

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 17]
```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
...
+    publishInfo.pPayload = "";
+    publishInfo.payloadLength = 0;
+    status = MQTT_GetPublishPacketSize( &publishInfo, &remainingLength, &packetSize );
+    TEST_ASSERT_EQUAL( MQTTSuccess, status );
+}
+
+/**
+ * @brief Tests that MQTT_SerializePublish works as intended.
+ */
+void test_MQTT_SerializePublish( void )
```
Comment (leegeth) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:874]: There is a new function to serialize only header of publish packet MQTT_SerializePublishHeader. That function also need to be tested. Since this is import of the existing tests from v4_beta, I think it can be considered in a separate PR.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 16 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:974]: Good point, I'll rearrange this so both QoS 0 and 2 are used. I don't think success case for QoS 1 is necessary since it's treated identically to QoS 2

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 1 above)
Comment (leegeth) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:1384]: QoS2

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 8 above)
Comment (leegeth) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:442]: Sounds good to me.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 15 above)
Comment (leegeth) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:868]: Yes. I think that test belong here.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 7 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:428]: Good catch, it seems this was misplaced in v4_beta and copied over to here. I'll move the test and delete it here

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 10 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:472]: Removed test since NULL payload should be allowed

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 11 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:484]: Added

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 6 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:419]: Added

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 12 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:523]: I think removing these will be best, I left these here so other reviewers could chime in. Will remove these after another round of review

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 18]
```diff
@@ -1267,9 +1267,12 @@ static void serializeConnectPacket( const MQTTConnectInfo_t * const pConnectInfo
         pIndex = encodeString( pIndex,
                                pWillInfo->pTopicName,
                                pWillInfo->topicNameLength );
-        pIndex = encodeString( pIndex,
-                               pWillInfo->pPayload,
-                               ( uint16_t ) pWillInfo->payloadLength );
+        if( pWillInfo->pPayload != NULL )
```
Comment (leegeth) [inline libraries/standard/mqtt/src/mqtt_lightweight.c:1270]: please add a comment here why we have to add this check instead of a parameter  validation.

Comment (muneebahmed10) [reply libraries/standard/mqtt/src/mqtt_lightweight.c]: Now that I think about it, this is incorrect since the length of the payload needs to be encoded in the next two bytes. Moved the check to inside `encodeString`, before the `memcpy` is called.

Comment (leegeth) [reply libraries/standard/mqtt/src/mqtt_lightweight.c]: Yes. That is true. Length of 0 has to be added. We just need to skip the memcpy.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 17 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:870]: Handled in #968

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c (hunk 8 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:456]: Verification of serialized buffers added in #968

Comment (leegeth) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:456]: Looks good to me.

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 19]
```diff
@@ -235,6 +323,1151 @@ static size_t encodeString( uint8_t * pDestination,
...
+
+/**
+ * @brief Tests that MQTT_SerializeConnect works as intended.
+ */
+void test_MQTT_SerializeConnect( void )
+{
+    MQTTConnectInfo_t connectInfo;
+    MQTTPublishInfo_t willInfo;
+    size_t remainingLength = 0;
+    uint8_t buffer[ 70 ];
```
Comment (gkwicker) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:429]: One thing I like to do in unit tests like this is to add extra space at the beginning and end of any buffer which will be written by the function under test, and fill those spaces with special values.  Then you execute the function, giving it an address just after your special values and verify afterwards that your special values are still in place.  This is particularly important when the function writing the buffer depend on some passed in length value which could be bogus (e.g. -1).

I usually create some utility functions to do this and use them throughout whenever a buffer is getting formatted.

Comment (muneebahmed10) [reply libraries/standard/mqtt/utest/mqtt_lightweight_utest.c]: Added

File: libraries/standard/mqtt/utest/mqtt_lightweight_utest.c [hunk 20]
```diff
@@ -323,6 +333,45 @@ static size_t encodeString( uint8_t * pDestination,
...
+ * @param[in] bufferLength Total length of buffer.
+ */
+static void padAndResetBuffer( uint8_t * pBuffer, size_t bufferLength )
+{
+    int i = 0;
+
+    for( i = 0; i < BUFFER_PADDING_LENGTH; i++ )
+    {
+        pBuffer[ i ] = BUFFER_PADDING_BYTE;
+        pBuffer[ bufferLength - 1 - i ] = BUFFER_PADDING_BYTE;
```
Comment (gkwicker) [inline libraries/standard/mqtt/utest/mqtt_lightweight_utest.c:350]: nice
